### PR TITLE
fix: Overview page Ended challenges should not be displayed in challenge portlet of overview page - MEED-1654

### PR DESCRIPTION
**Pull request by AzmiTouil**
Prior to this change, ended challenges are displayed in the challenge portlet

### CODE REVIEW

File: services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java [hunk 1]
```diff
@@ -148,32 +149,35 @@
...
 )
 @NamedQuery(
     name = "GamificationActionsHistory.findMostRealizedRuleIds",
-    query = "SELECT a.ruleId FROM GamificationActionsHistory a where a.type= :type AND (a.domainEntity.audienceId in (:spacesIds) OR a.domainEntity.audienceId = null ) AND a.createdDate >= :fromDate AND a.createdDate < :toDate " +
-            "group by a.ruleId order by count(*) DESC"
+    query = "SELECT a.ruleEntity.id FROM GamificationActionsHistory a where a.type= :type" +
+            " AND (a.domainEntity.audienceId in (:spacesIds) OR a.domainEntity.audienceId = null )" +
+            " AND a.createdDate >= :fromDate AND a.createdDate < :toDate" +
+            " AND a.ruleEntity.startDate <= :nowDate AND a.ruleEntity.endDate >= :nowDate" +
+            " group by a.ruleEntity.id order by count(*) DESC"
```
Comment (boubaker) [inline services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java]: Now, that we upgraded to JPA 2.2, we can use in JPQL `JOIN ... ON ...` for better performances and code clearness.
In addition using `= null` isn't recommended, we should use `IS NULL`.

```suggestion
    query = "SELECT r.id FROM GamificationActionsHistory a" +
        " JOIN a.ruleEntity r " +
        " ON  (r.startDate IS NULL OR r.startDate <= :nowDate)" +
        " AND (r.endDate IS NULL OR r.endDate >= :nowDate)" +
        " JOIN a.domainEntity d " +
        " ON d.audienceId IS NULL OR d.audienceId IN (:spacesIds)" +
        " WHERE a.type= :type" +
        " AND a.createdDate >= :fromDate AND a.createdDate < :toDate" +
        " group by r.id order by count(*) DESC"
```

File: services/src/test/java/org/exoplatform/addons/gamification/service/AnnouncementServiceTest.java [hunk 2]
```diff
@@ -18,6 +18,7 @@
 package org.exoplatform.addons.gamification.service;
 
 import org.exoplatform.addons.gamification.IdentityType;
+import org.exoplatform.addons.gamification.entities.domain.configuration.RuleEntity;
```
Comment (boubaker) [inline services/src/test/java/org/exoplatform/addons/gamification/service/AnnouncementServiceTest.java]: useless import ?

File: services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java [hunk 3]
```diff
@@ -28,6 +28,7 @@
 import javax.ws.rs.core.MultivaluedMap;
 import javax.ws.rs.core.SecurityContext;
 
+import org.exoplatform.addons.gamification.entities.domain.configuration.RuleEntity;
```
Comment (boubaker) [inline services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java]: useless import ?

File: services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java [hunk 4]
```diff
@@ -194,7 +195,7 @@ public void testGetAllAnnouncementByChallenge() throws Exception {
     Challenge challenge = new Challenge(0,
                                         "update challenge",
                                         "challenge description",
-                                        1l,
+                                        1L,
```
Comment (boubaker) [inline services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java:197]: just by curiosity and for me to learn, why this change ?

File: services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java (hunk 1 above)
Comment (AzmiTouil) [reply services/src/main/java/org/exoplatform/addons/gamification/entities/domain/effective/GamificationActionsHistory.java]: Done!

File: services/src/test/java/org/exoplatform/addons/gamification/service/AnnouncementServiceTest.java (hunk 2 above)
Comment (AzmiTouil) [reply services/src/test/java/org/exoplatform/addons/gamification/service/AnnouncementServiceTest.java]: removed!

File: services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java (hunk 3 above)
Comment (AzmiTouil) [reply services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java]: removed!

File: services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java (hunk 4 above)
Comment (AzmiTouil) [reply services/src/test/java/org/exoplatform/addons/gamification/rest/TestAnnouncementRest.java:197]: No practical difference, it's just to remove the editor's warning and eliminate the possibility of confusion between a lowercase **l** and the number **1**

**Review by boubaker — APPROVED** [review]
